### PR TITLE
BIP 143: Transaction signature verification for version 0 and version 1 witness program

### PR DESCRIPTION
**Pull request by jl2012**
This proposal defines a new transaction digest algorithm for signature verification in version 0 and version 1 witness program, in order to minimize redundant data hashing in verification, and to cover the input value by the signature.

### CODE REVIEW

File: bip-segwit-checksig.mediawiki [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  // Version
+  ss << txTo.nVersion;
+  // Input prevouts/nSequence (none/all, depending on flags)
+  ss << hashPrevouts;
+  ss << hashSequence;
+  // The input being signed (replacing the scriptSig with scriptCode + amount)
+  // The prevout may already be contained in hashPrevout, and the nSequence
+  // may already be contain in hashSequence.
+  ss << txTo.vin[nIn].prevout;
+  ss << static_cast<const CScriptBase&>(scriptCode);
```
Comment (NicolasDorier) [inline bip-segwit-checksig.mediawiki]: render as 
```
ss << static_cast&lt;const&gt;&lt;/const&gt;(scriptCode);
```

**Comment by NicolasDorier** [comment]
So, this basically redefines the 4 ECDSA OPs in the case of segwit scripts.

I am mixed about it, signature checkers already know the transaction being spent. From that they can parse the input being signed, and see if it is segwit. If it is, then change the hash rule, this is easy change.

However, I am a bit concerned about needing the previous input values. Actually transaction checkers don't have this information. Requiring it will make lots more code to change to support segwit, because the signature checker need now more data than before. It might not really be a valid complain though considering the upside.

**Comment by sipa** [comment]
On Jan 6, 2016 01:37, "Nicolas Dorier" <notifications@github.com> wrote:

> However, I am a bit concerned about needing the previous input values.
Actually transaction checkers don't have this information. Requiring it
will make lots more code to change to support segwit, because the signature
checker need now more data than before. It might not really be a valid
complain though considering the upside.

I am a bit confused what signature checker you are talking about. Either
you are a full node that maintains a UTXO set, and you have the amounts, or
you can't usefully verify anything at all. If you're talking about the
signing code: that should have access to amounts or it couldn't construct a
correct transaction.

However, I don't consider signing amounts or not a deal breaker. We had to
propose changing the sighashing algorithm to fix the O(n^2) behaviour, and
if we're doing that, it would seem stupid to not also make the probably
most-often requested related feature (amount signing). If there is
substantial controversy about it, I think we can remove that, and postpone
it to be an optional feature of a future softfork.

**Comment by NicolasDorier** [comment]
I don't think this will be controversial.
My point is that the method (in all frameworks and language)

    EvalScript(Script scriptPubKey, Script scriptSig, Transaction, Input)

Will become

    EvalScript(TxIn previous, Script scriptSig, Transaction, Input)

And same thing for the "Sign" method.
In practice, this will be easy to fix, people who can fetch the scriptPubKey before will most likely be able to fetch the TxIn easily as you said. But this will not be automatic.

Frankly, I don't think that matter, wallet providers, hardware wallet provider, and service providers like myself want so much the benefit of having the input value in the hash that nobody will likely complain.

File: bip-segwit-checksig.mediawiki (hunk 1 above)
Comment (jl2012) [inline bip-segwit-checksig.mediawiki]: It seems a bug of github. Any idea?

Comment (btcdrak) [inline bip-segwit-checksig.mediawiki]: yeah looks like output escaping.

Comment (MarcoFalke) [inline bip-segwit-checksig.mediawiki]: Not a bug. There is a syntax error in your mediawiki code: https://github.com/jl2012/bips/pull/5

Also you can activate code highlighting on the cpp code.

**Comment by FelixWeis** [comment]
This update seems like the perfect opportunity to get rid of that CHECKMULTISIG bug where one item too many is popped off the stack.

**Comment by jl2012** [comment]
@FelixWeis I'm not sure if it's worth extra testing and reviewing. The improvement is just 1 byte per CHECKMULTISIG. And CHECKMULTISIG will become obsolete when we introduce new signature scheme like Schnorr after the segwit softfork.

**Comment by FelixWeis** [comment]
If schnorr is replacing these completly not fixing makes sense, thanks!

**Pull request merged**